### DISCUSSION TITLE
Inserting data into Supabase using Azure Function (python)

### DISCUSSION THREAD

**Discussion by fertin** [supabase/supabase#29279] [Questions]
Hi all,

I have an azure function which processes PDFs, Images and other unstructured data sources through a blob trigger. The output is a json which is uploaded to a blob storage as well as in the end of the function being inserted into a table in Supabase.

The function runs from end to end if the Supabase upload is commented out but otherwise it does not execute any of code. The json-output is not uploaded to blob storage, even though the supabase insertion is executed after the blob upload.

The environment variables are defined and Azure Functions recognizes the function to be without any errors, so the issue should not be within Azure. I have not set any IP limitations in Supabase either.

For authentication I use an admin key, and the whole script runs perfectly if I run it locally using Azure Functions Core Tools. The files are processed and the output inserted into Supabase as expected.

Have any of you experienced something similar? I have assumed the issue was within Azure, but I have not found the issue, so now I am looking towards Supabase for e.g. blocking the connection/upload, but I have not found documentation or alike which could explain it.

I hope you have some guidance. :)

Best,

Victor

**Comment by silentworks** [comment]
When asking questions related to code please show code. It's hard to say what could be going wrong without seeing any code.

**Comment by fertin** [comment]
Sorry for that. Here is the related code. (I have attached an image of it as well)

The input is of json-format and is validated against the table structure in Supabase.

```
import pandas as pd
from supabase import create_client
import os

url = os.getenv("supabase_url")
key = os.getenv("supabase_admin_key")
supabase = create_client(url, key)

def fix_date(date):
    try:
        date = pd.to_datetime(date, format="mixed")
        return date.strftime('%Y-%m-%d')
    except:
        return None

# # Insert invoice metadata
def add_invoice_data(data, CVR):
    new_data = {"CVR": CVR,
                "invoiceDate": fix_date(data["invoiceDate"]),
                "startDate": fix_date(data["StartDate"]),
                "endDate": fix_date(data["EndDate"]),
                "supplierName": data["supplier"],
                "filename": data["Filename"],
                "documentCategory": data["documentCategory"],
                "supplierCVR": data["cvr"],
                "BFE": data["BFE"],
                "url": data["url"]}
    response = supabase.table('invoice_file').insert([new_data]).execute()
    return response.data[0]["id"]


# # Insert line items
def add_lineItem_data(data, file_id):
    # # Prepare the list of line items for batch insertion
    line_items = []
    for data_temp in data["listData"]:
        new_data = {
            "filenameID": file_id,
            "consumptionDate": fix_date(data_temp.get("consumptionDate")),
            "item": data_temp.get("item"),
            "amount": data_temp.get("amount"),
            "unit": data_temp.get("unit"),
            "spend": data_temp.get("spend"),
            "currency": data_temp.get("currency"),
            "emmisionFactorID": data_temp.get("Category"),
            "pageNumber": data_temp.get("page")
        }
        line_items.append(new_data)

    # # Perform a single batch insert
    if line_items:
        response = supabase.table('invoice_lineitem').insert(line_items).execute()

# # Run all
def insert_file_data(data, CVR):
    try:
        file_id = add_invoice_data(data, CVR)
        add_lineItem_data(data, file_id)
        print("Successfully uploaded data")
    except Exception as e:
        print(e)
```
